### PR TITLE
Updating RNN to include all types PyTorch has, and uses CUDNN

### PR DESCRIPTION
**Pull request by ebetica**
Changes:

- Separate initialize_containers and initialize_parameters to show correct usage. If you don't, then the options framework won't work, since sometimes container construction requires options to be set (`LSTM(2, 2).nlayers().make()`). This implies the correct usage more.
- Added GRUs, LSTMs, and RNNs with a tanh or relu nonlinearity
- Tests LSTMs against pytorch implementation, and tests all against a stupid counting example (number of 1s in a string of 5 binary numbers)

Unimplemented:
- Bidirectional
- Batch first
- Fused non-CUDNN CUDA kernels (not bound in PyTorch ATen)
- PackedSequence

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -42,7 +42,8 @@ AUTOGRAD_CONTAINER_CLASS(MyModel) {
 ```
 
 Some things are not implemented:
-- Batchnorm
-- SGD, Adagrad, RMSprop, and Adam are implemented: the rest of the optimizers are just copying Python code from PyTorch over.
+- SGD, Adagrad, RMSprop, and Adam are the only optimizers implemented
```
Comment (eugene-kharitonov) [inline README.md]: amsgrad also? ;)
